### PR TITLE
Additions, refinement and bugfixes

### PR DESCRIPTION
**Pull request by LindyHopperGT**
- Fixed a bug in CallOwnerFunction where it would not bind correctly.
- Fixed an bug where flow nodes were in the palette, even if they were in inaccessible plugins (now FlowNode palette respects plugin access rules)
- Added expanded CustomOutput access
- Clarified editor-accessible vs. runtime accessible information for custom in/outputs
- Split out some functionality for subsystem registration to be called separately (to support our object pooling model)

### CODE REVIEW

File: Source/Flow/Private/FlowAsset.cpp [hunk 1]
```diff
@@ -71,16 +76,45 @@ void UFlowAsset::PostDuplicate(bool bDuplicateForPIE)
 	}
 }
 
+void UFlowAsset::PostLoad()
+{
+	Super::PostLoad();
+
+	// If we removed or moved a flow node blueprint (and there is no redirector) we might loose the reference to it resulting
```
Comment (LindyHopperGT) [inline Source/Flow/Private/FlowAsset.cpp:83]: This code fixes-up flow assets that were corrupted by the fixed bug in CallOwnerFunction.

File: Source/Flow/Private/FlowAsset.cpp [hunk 2]
```diff
@@ -71,16 +76,45 @@ void UFlowAsset::PostDuplicate(bool bDuplicateForPIE)
...
 {
 	// validate nodes
 	for (const TPair<FGuid, UFlowNode*>& Node : Nodes)
 	{
-		if (Node.Value)
+		if (IsValid(Node.Value))
 		{
-			if (!IsNodeClassAllowed(Node.Value->GetClass()))
+			FText FailureReason;
+			if (!IsNodeClassAllowed(Node.Value->GetClass(), &FailureReason))
```
Comment (LindyHopperGT) [inline Source/Flow/Private/FlowAsset.cpp:111]: Validate will now report inaccessible flow nodes being used in a flow graph (that is, they're in an inaccessible plugin).

File: Source/Flow/Private/FlowAsset.cpp [hunk 3]
```diff
@@ -90,74 +124,124 @@ EDataValidationResult UFlowAsset::ValidateAsset(FFlowMessageLog& MessageLog)
...
-bool UFlowAsset::IsNodeClassAllowed(const UClass* FlowNodeClass) const
+bool UFlowAsset::IsNodeClassAllowed(const UClass* FlowNodeClass, FText* OutOptionalFailureReason) const
 {
-	if (FlowNodeClass == nullptr)
+	if (!IsValid(FlowNodeClass))
+	{
+		return false;
+	}
+
+	if (!CanFlowNodeClassBeUsedByFlowAsset(*FlowNodeClass))
```
Comment (LindyHopperGT) [inline Source/Flow/Private/FlowAsset.cpp:144]: We broke-out these checks into their own subfunctions (below) and added a new one to check plugin access rules for the node, relative to this flow graph.

File: Source/Flow/Private/FlowAsset.cpp [hunk 4]
```diff
@@ -90,74 +124,124 @@ EDataValidationResult UFlowAsset::ValidateAsset(FFlowMessageLog& MessageLog)
...
+	if (!CanFlowAssetReferenceFlowNode(*FlowNodeClass, OutOptionalFailureReason))
 	{
 		return false;
 	}
 
-	UFlowNode* NodeDefaults = FlowNodeClass->GetDefaultObject<UFlowNode>();
+	return true;
+}
+
+bool UFlowAsset::CanFlowNodeClassBeUsedByFlowAsset(const UClass& FlowNodeClass) const
```
Comment (LindyHopperGT) [inline Source/Flow/Private/FlowAsset.cpp:163]: We split-out some of the pre-existing checks into single-topic functions to improve code clarity.

File: Source/Flow/Private/FlowAsset.cpp [hunk 5]
```diff
@@ -341,6 +425,56 @@ UFlowNode_CustomInput* UFlowAsset::TryFindCustomInputNodeByEventName(const FName
 	return nullptr;
 }
 
+UFlowNode_CustomOutput* UFlowAsset::TryFindCustomOutputNodeByEventName(const FName& EventName) const
```
Comment (LindyHopperGT) [inline Source/Flow/Private/FlowAsset.cpp:428]: Access function for custom output by event name.

File: Source/Flow/Private/FlowAsset.cpp [hunk 6]
```diff
@@ -341,6 +425,56 @@ UFlowNode_CustomInput* UFlowAsset::TryFindCustomInputNodeByEventName(const FName
...
+			{
+				return CustomOutput;
+			}
+		}
+	}
+
+	return nullptr;
+}
+
+TArray<FName> UFlowAsset::GatherCustomInputNodeEventNames() const
```
Comment (LindyHopperGT) [inline Source/Flow/Private/FlowAsset.cpp:444]: The "name" versions of the custom input/output are editor-only, so if the names of inputs/outputs are used at runtime, we need to gather them from the nodes list.  These functions handle that work.

File: Source/Flow/Private/FlowComponent.cpp [hunk 7]
```diff
@@ -44,6 +44,11 @@ void UFlowComponent::BeginPlay()
 {
 	Super::BeginPlay();
 
+	RegisterWithFlowSubsystem();
```
Comment (LindyHopperGT) [inline Source/Flow/Private/FlowComponent.cpp:47]: Our object pooling model required different timing for registration with the subsystem than strictly in Begin/End Play, so we split these parts out into dedicated functions

File: Source/Flow/Private/Nodes/World/FlowNode_CallOwnerFunction.cpp [hunk 8]
```diff
@@ -368,27 +387,20 @@ UClass* UFlowNode_CallOwnerFunction::GetParamsClassForFunctionName(const UClass&
...
 		return false;
 	}
 
 	const UClass* ExistingParamsClass = GetExistingParamsClass();
+	UClass* RequiredParamsClass = GetRequiredParamsClass();
 
 	const bool bNeedsAllocateParams = 
 		!IsValid(ExistingParamsClass) ||
-		!ExistingParamsClass->IsChildOf(RequiredParamsClass);
+		ExistingParamsClass != RequiredParamsClass;
```
Comment (LindyHopperGT) [inline Source/Flow/Private/Nodes/World/FlowNode_CallOwnerFunction.cpp:403]: Requiring explicit class match now.  This is required by the new approach for allocating the correct params object for a CallOwner function.

File: Source/Flow/Private/Nodes/World/FlowNode_CallOwnerFunction.cpp [hunk 9]
```diff
@@ -97,12 +97,31 @@ void UFlowNode_CallOwnerFunction::PostLoad()
...
+	//  (it could be removed in a future release, once all assets have been updated)
+	if (ParamsProperty->PropertyClass != UFlowOwnerFunctionParams::StaticClass())
+	{
+		ParamsProperty->SetPropertyClass(UFlowOwnerFunctionParams::StaticClass());
+	}
+}
+
+bool UFlowNode_CallOwnerFunction::CanEditChange(const FProperty* InProperty) const
+{
+	if (!Super::CanEditChange(InProperty))
```
Comment (LindyHopperGT) [inline Source/Flow/Private/Nodes/World/FlowNode_CallOwnerFunction.cpp:112]: We're not allowing the user to choose the params object directly any longer (the way this was done previously was dangerous).  So now the params object is only created when selecting the function, and this CanEditChange prevents the user from choosing an incompatible params object.

File: Source/Flow/Public/FlowAsset.h [hunk 10]
```diff
@@ -210,6 +206,25 @@ class FLOW_API UFlowAsset : public UObject
 		}
 	}
 
+public:
```
Comment (LindyHopperGT) [inline Source/Flow/Public/FlowAsset.h:209]: Regrouped and expanded the custom input/output access functions

File: Source/FlowEditor/Private/Asset/FlowAssetToolbar.cpp [hunk 11]
```diff
@@ -202,6 +202,7 @@ void FFlowAssetToolbar::BuildAssetToolbar(UToolMenu* ToolbarMenu) const
 		// add buttons
 		Section.AddEntry(FToolMenuEntry::InitToolBarButton(FFlowToolbarCommands::Get().RefreshAsset));
 		Section.AddEntry(FToolMenuEntry::InitToolBarButton(FFlowToolbarCommands::Get().ValidateAsset));
+		Section.AddEntry(FToolMenuEntry::InitToolBarButton(FFlowToolbarCommands::Get().EditAssetDefaults));
```
Comment (LindyHopperGT) [inline Source/FlowEditor/Private/Asset/FlowAssetToolbar.cpp:205]: Moved this section to align with the others

File: Source/FlowEditor/Private/DetailCustomizations/FlowOwnerFunctionRefCustomization.cpp [hunk 12]
```diff
@@ -69,8 +69,19 @@ TArray<FName> FFlowOwnerFunctionRefCustomization::GetFlowOwnerFunctionRefs(
 	TArray<FName> ValidFunctionNames;
 
 	// Gather a list of potential functions
-	TArray<FName> PotentialFunctionNames;
-	ExpectedOwnerClass.GenerateFunctionList(PotentialFunctionNames);
+	TSet<FName> PotentialFunctionNames;
+
+	const UClass* CurClass = &ExpectedOwnerClass;
```
Comment (LindyHopperGT) [inline Source/FlowEditor/Private/DetailCustomizations/FlowOwnerFunctionRefCustomization.cpp:74]: Needed to gather functions from superclasses, and not just the leaf class.

File: Source/FlowEditor/Private/Graph/FlowGraph.cpp [hunk 13]
```diff
@@ -31,7 +31,13 @@ UFlowGraph::UFlowGraph(const FObjectInitializer& ObjectInitializer)
 
 UEdGraph* UFlowGraph::CreateGraph(UFlowAsset* InFlowAsset)
 {
-	UEdGraph* NewGraph = CastChecked<UFlowGraph>(FBlueprintEditorUtils::CreateNewGraph(InFlowAsset, NAME_None, StaticClass(), UFlowGraphSchema::StaticClass()));
+	return CreateGraph(InFlowAsset, UFlowGraphSchema::StaticClass());
+}
+
+UEdGraph* UFlowGraph::CreateGraph(UFlowAsset* InFlowAsset, TSubclassOf<UFlowGraphSchema> FlowSchema)
```
Comment (LindyHopperGT) [inline Source/FlowEditor/Private/Graph/FlowGraph.cpp:37]: We use our own Schema (subclass) so we needed a CreateGraph function that accepted a user-specified schema.

We may want to remove the other version of this function (w/o the explicit schema param), but I didn't make that change in this changelist.

File: Source/FlowEditor/Private/Graph/FlowGraphSchema.cpp [hunk 14]
```diff
@@ -54,9 +54,9 @@ void UFlowGraphSchema::SubscribeToAssetChanges()
 	}
 }
 
-void UFlowGraphSchema::GetPaletteActions(FGraphActionMenuBuilder& ActionMenuBuilder, const UClass* AssetClass, const FString& CategoryName)
+void UFlowGraphSchema::GetPaletteActions(FGraphActionMenuBuilder& ActionMenuBuilder, const UFlowAsset* EditedFlowAsset, const FString& CategoryName)
```
Comment (LindyHopperGT) [inline Source/FlowEditor/Private/Graph/FlowGraphSchema.cpp:57]: This is reworked to not use the default flow asset, but instead use the actual flow asset (that is being edited) so we can ask the right class about which nodes are legal for the palette.

File: Source/FlowEditor/Public/DetailCustomizations/FlowOwnerFunctionRefCustomization.h [hunk 15]
```diff
@@ -48,6 +48,4 @@ class FFlowOwnerFunctionRefCustomization : public IFlowCuratedNamePropertyCustom
 	static TArray<FName> GetFlowOwnerFunctionRefs(const UFlowNode_CallOwnerFunction& FlowNodeOwner, const UClass& ExpectedOwnerClass);
 
 	static bool IsFunctionUsable(const UFunction& Function, const UFlowNode_CallOwnerFunction& FlowNodeOwner);
-	static bool DoesFunctionHaveExpectedParamType(const UFunction& Function, const UFlowNode_CallOwnerFunction& FlowNodeOwner);
```
Comment (LindyHopperGT) [inline Source/FlowEditor/Public/DetailCustomizations/FlowOwnerFunctionRefCustomization.h:51]: the fixed model for CallOwnerFunction configuration does not need this function any longer.

File: Source/Flow/Private/FlowAsset.cpp [hunk 16]
```diff
@@ -90,74 +124,124 @@ EDataValidationResult UFlowAsset::ValidateAsset(FFlowMessageLog& MessageLog)
...
+		if (!bAllowedInAsset)
+		{
+			return false;
+		}
+	}
+
+	return true;
+}
+
+bool UFlowAsset::CanFlowAssetReferenceFlowNode(const UClass& FlowNodeClass, FText* OutOptionalFailureReason) const
```
Comment (LindyHopperGT) [inline Source/Flow/Private/FlowAsset.cpp:227]: This is the new check if the node is accessible from the plugin context of the flow asset.

**Comment by MothDoctor** [comment]
Merged that, thanks a lot again! :)

**Pull request merged**